### PR TITLE
Couch to SQL Repeat Records PR 1: Write to SQL

### PR DESCRIPTION
**Pull request by millerdev**
This is the [first PR in the Couch to SQL migration process](https://commcare-hq.readthedocs.io/couch_to_sql_models.html#pr-1-add-sql-model-and-migration-management-command-write-to-sql) of Repeat Records. Commits for PR 2 (and eventually PR 3) can be found on [dm/sql-repeatrecord](https://github.com/dimagi/commcare-hq/compare/dm/sql-repeatrecord).

:tropical_fish: Review by commit.

https://dimagi-dev.atlassian.net/browse/SAAS-14299

### Safety story

Automated tests have been added where necessary to cover new functionality and to verify that Couch Repeat Records are synced to SQL as expected. In general, TDD was used to develop this PR. Repeater functionality should be unchanged.

### Automated test coverage

Yes.

### QA Plan

https://dimagi-dev.atlassian.net/browse/QA-6025

### Rollback instructions

A decision will need to be made about what to do with data in `SQLRepeatRecord` and `SQLRepeatRecordAttempt` tables If this PR is reverted. Otherwise, this PR can be reverted after deploy with no further considerations.

### CODE REVIEW

File: corehq/motech/repeaters/tests/test_dbaccessors.py [hunk 1]
```diff
@@ -50,7 +67,6 @@ def setUpClass(cls):
             domain=cls.domain,
             succeeded=True,
             repeater_id=cls.repeater_id,
-            next_check=before,
```
Comment (minhaminha) [inline corehq/motech/repeaters/tests/test_dbaccessors.py:53]: Why was this line removed/how was it breaking the tests?

File: corehq/apps/data_interfaces/tests/test_utils.py [hunk 2]
```diff
@@ -542,14 +542,10 @@ def create_repeat_records(cls):
...
             cls.couch_records.append(couch_record)
-
-            cls.sql_records.append(SQLRepeatRecord.objects.create(
-                domain=DOMAIN,
-                couch_id=couch_record._id,
-                payload_id=cls.instance_id,
-                repeater=cls.repeater,
-                registered_at=now,
-            ))
+        cls.sql_records = list(SQLRepeatRecord.objects.filter(
```
Comment (minhaminha) [inline corehq/apps/data_interfaces/tests/test_utils.py:545]: nit: You can get rid of line 534 (cls.sql_records = [])

File: corehq/motech/repeaters/tests/test_dbaccessors.py (hunk 1 above)
Comment (millerdev) [reply corehq/motech/repeaters/tests/test_dbaccessors.py:53]: It is not valid to have a `next_check` date in SQL when `state == Success`.

https://github.com/dimagi/commcare-hq/blob/f8abe9215410b3af3561296a5aca8b2b6687d58f/corehq/motech/repeaters/models.py#L1366-L1373

File: corehq/apps/data_interfaces/tests/test_utils.py (hunk 2 above)
Comment (millerdev) [reply corehq/apps/data_interfaces/tests/test_utils.py:545]: Thanks. This gets refactored in a [later commit](https://github.com/dimagi/commcare-hq/commit/3cd0bece1c890fd210b2ea2bc4f822aac5db3cce) (it will be in the next PR).

**Review by AmitPhulera — APPROVED** [review]
Looking good overall. Just a minor comment on tests.

I am assuming this is already in process of QA and is deployed on staging?

File: corehq/motech/repeaters/models.py [hunk 3]
```diff
@@ -909,6 +910,49 @@ def datasource_is_subscribed_to(domain, data_source_id):
...
+        self.failure_reason = ""
+    elif value == RECORD_CANCELLED_STATE:
+        self.succeeded = False
+        self.cancelled = True
+        self.failure_reason = ""
+    elif value == RECORD_FAILURE_STATE:
+        self.succeeded = False
+        self.cancelled = False
+        try:
+            reason = self.failure_reason
```
[the remark this answers is not in the document]
Comment (millerdev) [reply corehq/motech/repeaters/models.py:944]: jsonobject's [`JsonProperty` raises `AssertionError`](https://github.com/dimagi/jsonobject/blob/61fed25f1dbe9bf231ca2897d1b80b4dfee615b1/jsonobject/base_properties.pyx#L73) if the record does not have a `failure_reason`. That's annoying. It should raise `AttributeError`, which would allow standard use of `getattr`.

File: corehq/motech/repeaters/models.py [hunk 4]
```diff
@@ -1355,12 +1355,18 @@ def add_success_attempt(self, response):
         ``response`` can be a Requests response instance, or True if the
         payload did not result in an API call.
         """
+        # NOTE a 204 status here could mean either
+        # 1. the request was not sent, in which case response is
+        #    probably RepeaterResponse(204, "No Content")
+        # 2. the request was sent, and the remote end responded
```
[the remark this answers is not in the document]
Comment (millerdev) [reply corehq/motech/repeaters/models.py:1415]: Possibly if we want to keep it this way. I think a better long-term fix would be to change the way we internally signal that the request was not sent so there is no ambiguity.

**Comment by millerdev** [comment]
> I am assuming this is already in process of QA and is deployed on staging?

Correct.

**Review by dannyroberts — APPROVED** [review]
Looks great. For parts I couldn't follow super well, I looked for tests—and the way you paired small test changes with each commit made that easy!

File: corehq/motech/repeaters/models.py [hunk 5]
```diff
@@ -1049,8 +1049,7 @@ def wrap(cls, data):
             )]
         return self
 
-    @property
-    @memoized
+    @cached_property
```
[the remark this answers is not in the document]
Comment (millerdev) [reply corehq/motech/repeaters/models.py:1076]: Yes, not disqualifying.

File: corehq/motech/repeaters/management/commands/populate_repeatrecords.py [hunk 6]
```diff
@@ -78,31 +78,51 @@ def get_ids_to_ignore(self, docs):
...
+    def _get_couch_doc_count_for_domains(self, domains):
+        def count_domain_docs(domain):
+            return count_docs(startkey=[domain], endkey=[domain, {}])
+        return sum(count_domain_docs(d) for d in domains)
+
+    def _iter_couch_docs_for_domains(self, domains, chunk_size):
+        def iter_domain_docs(domain):
+            return iter_docs(chunk_size, startkey=[domain], endkey=[domain, {}])
+        for domain in domains:
+            yield from iter_domain_docs(domain)
```
[the remark this answers is not in the document]
Comment (millerdev) [reply corehq/motech/repeaters/management/commands/populate_repeatrecords.py:117]: They're called by the Couch to SQL migration framework code when migrating individual domains.

https://github.com/dimagi/commcare-hq/blob/bbe7d3577d5d61b0dcc314da35e8c1f1ed9d22a7/corehq/apps/cleanup/management/commands/populate_sql_model_from_couch_model.py#L368-L370

File: corehq/motech/repeaters/tests/test_couchsqlmigration.py [hunk 7]
```diff
@@ -250,6 +251,18 @@ def test_migration_with_repeater_added_after_start(self):
...
+            docs = {}
+            for repeater in [self.repeater1, self.repeater2]:
+                doc, obj = self.create_repeat_record(unwrap_doc=False, repeater=repeater)
+                doc.save(sync_to_sql=False)
+                docs[repeater.domain] = doc
+
+            call_command('populate_repeatrecords', domains=["other"], log_path=log.path)
+            self.assertIn(f'Created model for RepeatRecord with id {docs["other"]._id}\n', log.content)
+            self.assertNotIn(docs["test"]._id, log.content)
+
```
[the remark this answers is not in the document]
Comment (millerdev) [reply corehq/motech/repeaters/tests/test_couchsqlmigration.py:411]: https://github.com/dimagi/commcare-hq/pull/33978/commits/d624c4adae79950b3b7040febbd4798daf906858

**Review by dannyroberts — APPROVED** [review]

File: corehq/motech/repeaters/models.py [hunk 8]
```diff
@@ -1038,6 +1039,14 @@ def _migration_get_submodels(cls):
     def _migration_get_sql_model_class(cls):
         return SQLRepeatRecord
 
+    def save(self, *args, sync_attempts=False, **kw):
+        with enable_attempts_sync_to_sql(self, sync_attempts):
+            return super().save(*args, **kw)
+
+    def _migration_sync_submodels_to_sql(self, sql_object):
+        if self._should_sync_attempts:
```
Comment (kaapstorm) [inline corehq/motech/repeaters/models.py:1047]: I didn't follow the paths of what calls `__migration_sync_to_sql()`, but I'm guessing that you are sure that this cannot be reached outside the `enable_attempts_sync_to_sql()` context manager?

**Review by kaapstorm — APPROVED** [review]

File: corehq/motech/repeaters/models.py (hunk 8 above)
Comment (millerdev) [reply corehq/motech/repeaters/models.py:1047]: Yes, this is only called within the context of `.save()`, which uses the context manager. Having the attribute not set outside the context manager will alert us (hopefully in tests, if not then in Sentry) if any code path arises that violates that precondition.

File: corehq/motech/repeaters/models.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (kaapstorm) [reply corehq/motech/repeaters/models.py:1415]: > Is it possible that they were all changed from `return True` to return `RepeaterResponse(204, "No content")`?

I checked. They were all changed:
```python
Repeater.send_request() -> requests.Response
Dhis2Repeater.send_request() -> requests.Response | RepeaterResponse
FHIRRepeater.send_request() -> requests.Response | RepeaterResponse
ReferCaseRepeater.send_request() -> requests.Response | RepeaterResponse
DataRegistryRepeater.send_request() -> requests.Response | RepeaterResponse
OpenmrsRepeater.send_request() -> RepeaterResponse
Dhis2EntityRepeater.send_request() -> RepeaterResponse
```

We should update that docstring, in a separate PR.

> a test asserts that a (SQL) record created with `response=True` results in `state == Success` (not `Empty`). Should that test be changed?

I guess that test can be removed in the PR that updates the docstring.

**Comment by millerdev** [comment]
Latest commits address data states discovered while running `populate_repeatrecords` on staging.

**Review by AmitPhulera — APPROVED** [review]

File: corehq/motech/repeaters/tests/test_couchsqlmigration.py [hunk 9]
```diff
@@ -0,0 +1,548 @@
...
+        doc, _ = self.create_repeat_record(unwrap_doc=False)
+        doc.save()  # sync to SQL, but do not save attempts
+        with templog() as log, patch.object(transaction, "atomic", atomic_check):
+            call_command('populate_repeatrecords', log_path=log.path)
+            self.assertNotIn('has differences:', log.content)
+        obj = SQLRepeatRecord.objects.get(couch_id=doc._id)
+        self.assertEqual(len(obj.attempts), len(doc.attempts))
+        self.assertTrue(obj.attempts)
+
+    def test_migrate_record_with_unynced_sql_attempts2(self):
```
[the remark this answers is not in the document]
Comment (millerdev) [reply corehq/motech/repeaters/tests/test_couchsqlmigration.py]: See https://github.com/dimagi/commcare-hq/pull/33978/commits/9e7d4d3a3cb2517189610de152ba504b37587bf4, in which I renamed the test and added a comment.

**Review by AmitPhulera — APPROVED** [review]
